### ISSUE TITLE
overrideMap generation can create cycles

### ISSUE THREAD

**Issue by goleary** [GoogleChromeLabs/text-fragments-polyfill#61]
I'm using these sweet utility functions & have noticed something that y'all might want to know about.

The process that creates an overrideMap & traverse nodes (to ensure block boundaries are detected) is nifty, but I've see it result in circular references resulting in an infinite loop.

Here you can see an example where the map contains a cycle:
![image](https://user-images.githubusercontent.com/16123225/118588553-73543880-b753-11eb-9aba-b1a38d8a6808.png)

What I have done is simply delete the entry for the current node in [forwardTraverse](https://github.com/GoogleChromeLabs/text-fragments-polyfill/blob/f0159ae64cc800d04fdbee6499f569a94ccf4915/src/fragment-generation-utils.js#L1017) once we map to it's parent resulting in progression past it's parent the second time around.
![image](https://user-images.githubusercontent.com/16123225/118588690-c9c17700-b753-11eb-9bb8-7e7f3be5cb79.png)
I'm not positive this results in the exact intended behavior, but it seems to be working well enough for me.

You can see for yourself at [this link](https://www.notboring.co/p/transforming-tata-0b9) with this selection:

![image](https://user-images.githubusercontent.com/16123225/118589039-71d74000-b754-11eb-94ff-3b5c120f8b96.png)

The [chrome extension](https://chrome.google.com/webstore/detail/link-to-text-fragment/pbcodcjpfjdpcineamnnmbkkmkdpajjg?hl=en) craps out for this selection, but with my change it manges to create a valid TextFragment.
![image](https://user-images.githubusercontent.com/16123225/118589342-0fcb0a80-b755-11eb-9eac-9771a4e3a1d2.png)

If my diff doesn't break any other assumptions I'd be happy to submit a PR.

**Comment by tomayac** [comment]
Thanks a lot for investingating. @goleary! @tfmar, could you look into this?

**Comment by tomayac** [comment]
@sebsg or @SebLalancette, while @tfmar is OoO, are you able to look into this?

**Comment by guillaumejenkins** [comment]
Just a quick update, I'm still investigating this, I want to make sure I fully understand the override map generation and its use so I can determine whether removing ancestor entries from the map after consumption is the right fix. Just a heads up that this week is a 3-day work week for me, so I might need until some time next week

**Comment by guillaumejenkins** [comment]
What a fascinating bug, it's actually 2 problems in one! `createForwardOverrideMap` generating a cycle is definitely a bug, but in this case the infinite loop is the symptom, not the cause, of the failure. What's truly causing the fragment generation to fail is #63, because we should be returning early but aren't, so it allows this infinite loop to take place.

In fact, I am not sure it is ever possible to hit this infinite loop bug in a legit scenario (assuming #63 is fixed), because we would always hit a block boundary (and thus, return) before entering the cycle. Of course I will still try to fix the `overrideMap` generation logic in a separate fix so we are extra safe, but it's an interesting thought.

Thanks for reporting and for the initial investigation!

**Comment by goleary** [comment]
Interesting, nice work getting to the bottom of it, sounds like quite the doozy!

**Comment by tomayac** [comment]
Can this Issue be closed now?

**Comment by guillaumejenkins** [comment]
I was thinking of using this issue as a tracking bug for the fact that the `overrideMap` can be generated with a cycle in it. But  the link generation failure in this specific example has been fixed, so I don't mind opening a separate issue for the cycle bug specifically.

**Comment by tomayac** [comment]
Let’s leave this Issue open then. Maybe update the title. I’ll hold on to releasing a new version until it’s fixed. Especially the App Store submission is a hassle.

**Comment by guillaumejenkins** [comment]
I figured out what the issue was: when creating an override, we were using the child node's `nextNode()` as the ancestor's override, when in fact we should have been using the child node's existing override (if any). It's easier to illustrate with an example, so take the following HTML:
```
<p><em>hello</em></p>
```

There are 3 nodes: `p`, `em`, and `text`. When generating the override map starting from `text`, we first set `text`'s override to `null`, since it doesn't have any successor:
```
text -> null
```

Next, we move to the parent, `em`. We find `em`'s child, `text`, so we change `text`'s override to point to `em` instead. At the same time, we add an override so that `em` continues the traversal to whatever was supposed to come after `text` (which is `null`). We get this:
```
text -> em
em -> null
```

We move on to `em`'s parent, `p`. We find its child, `em`, so we change `em`'s override to point to `p`. Now, as before, we also need to add an override so that `p` continues the traversal to whatever was supposed to come after `em`. When using `walker.nextNode()`, what comes after `em` is `text`. But we had previously overridden that `em` should instead go to `null`, not `text`. Thus, by using `walker.nextNode()`, the grandparent of a node will **always** create a cycle back to its grandchild, because it uses its child's next node, which ignores the fact that its child's next node had been overridden. So we end up with this cycle:
```
text -> em
em -> p
p -> text
```

The fix is to detect that `em`'s _true_ next node was the overriden `null`, not the walker's `nextNode()`. So we'll end up with this, which doesn't contain any cycle:
```
text -> em
em -> p
p -> null
```

See the linked pull request for the exact fix and unit test
